homeassistant/components/light/__init__.py

```diff
@@ -240,6 +239,14 @@ def preprocess_turn_on_alternatives(params):
     if brightness_pct is not None:
         params[ATTR_BRIGHTNESS] = int(255 * brightness_pct/100)
 
+    # Make color changes available as both XY and RGB for platforms
+    if ATTR_XY_COLOR in params:
+        color_rgb = color_util.color_xy_to_RGB(*params[ATTR_XY_COLOR])
+        params[ATTR_RGB_COLOR] = color_rgb
+    elif ATTR_RGB_COLOR in params:
+        color_xy = color_util.color_RGB_to_xy(*params[ATTR_RGB_COLOR])
+        params[ATTR_XY_COLOR] = color_xy
+
```

Comment:
This seems to cause inconsistent behavior when no brightness is provided:

* For light platforms preferring xy, the brightness component of RGB is lost.
* For light platforms preferring RGB, an arbitrary full brightness is added to xy calls.

For example, this breaks xy for the LIFX platform which relies on only one of xy or RGB being available and hence will trust the brightness information in RGB colors.

So setting an xy color on a group with both Hue and LIFX will cause one of them to go to full brightness but not the other.

Setting an RGB color on the same group will only make one of them change its brightness.

I do not have a solution for this. It was discussed in #6727 but consensus was not reached before it was closed.

Reply:
Those changes were intentional, although I'll have to double-check how LIFX looks for colors. Extracting brightness from the RGB color is a bad behavior that IMO we should eliminate. If the service call should change the brightness it should specify the brightness attribute. Otherwise the color should never be used to change the brightness. Right now lots of platforms have the weird behavior of clicking on the color wheel and having the brightness change.

Reply:
I'm not following.

Your current implementation _does_ change brightness on an RGB-based light platform.

Is that platform expected to notice that no `brightness` attribute is present, convert RGB to HSV, replace V with the current brightness and finally convert back to RGB?

Reply:
It obviously depends on how the platform is implemented, but lots of platforms see that they can extract a brightness from the RGB values, and assume that's what they should use, which is the bad behavior we want to eliminate. If a bulb is at 50% brightness, and the user clicks red on the color wheel, it shouldn't move to 100% brightness. Open to other ideas, but that's the end goal I'd like to achieve. The color attribute should only be used to change color.

Reply:
I see where you are coming from but I don't think I agree. I don't think it is bad behavior to honor the brightness in RGB, I think it is the right thing to do with RGB. Otherwise `[1,0,0]`, `[127,0,0]` and `[255,0,0]` all mean the same thing and that definitely seems like a recipe for confusion.

I do agree that RGB is a monster to handle in any intuitive way. Maybe we _can_ phase it out? If we add `hue` (0-360) and `saturation` (0-100) attributes we can express the same things as RGB can but with the individual attributes we gain the much-desired ability to specify just a color. I think we would also want an `hsv` shorthand to keep full color specifications succinct.

The migration would obviously be quite painful but I don't see that we would miss anything if it came through?

As for the color wheel, why doesn't the frontend simply plug in its current brightness in the conversion, rather than full brightness?

Reply:
Although neither behavior is ideal, I'd much rather have all three of those RGB examples do the same thing, than have RGB destructively change brightness with no reliable way to work around it. It's pretty easy to realize those do the same thing and that brightness needs to be included if that's the intention.

I'd be open to always transforming color requests to hue and saturation in our preprocess function, and have platforms deal with only that, but I don't think it's reasonable to remove RGB from the server API.

Trying to pick out the current brightness on the frontend side is a recipe for race conditions.  The frontend should only be making an API request for the actual intended change, so that the device can sort out what to do correctly. The color wheel is also just an example. All of the same arguments apply to a service call that only specifies a color.

Reply:
Why is it not reasonable to remove RGB? Because everybody is using it (I agree, that's the painful part I talked about) or because it is needed in some cases (I do not see that)?

I also agree on the race issue but for the color wheel specifically, I think that would be preferable for the short-term (like 0.61). Many platforms have such a race anyway because they can only apply a cached brightness when they have to present RGB to the light hardware.

Reply:
Ignoring the fact that removing RGB would be a painful breakage, it's such an ubiquitously understood format I'm not sure that we want to even if we could. I'd rather though just keep it around as an alternative color specification similar to brightness_pct as a convenience for our API. I don't think there's any use case that couldn't be solved otherwise though.

A good portion of our platforms already have a native color-only changing API supported by the device itself, and for those that don't I think we should push that calculation as close to the device as possible.

I'd also be open to a short-term fix if someone wants to put it together, but that could possibly come with it's own side-effects and new bugs.

Reply:
Okay, I understand your position now.

However, I don't think the proposed RGB-minus-brightness is ubiquitously understood. Nowhere else do you have to specify an explicit brightness for an RGB color.

Note that changing the meaning of RGB to RGB-minus-brightness would also be a huge breaking change as everybody would have to add that explicit brightness to all colors. We probably couldn't even help with deprecation warnings because the configuration would not be deprecated, just changed.

So, ignoring the pain, I do think the best way would be to fully remove RGB along with all the confusion it causes. Though, if RGB only acted as a preprocessed alias, I wouldn't hate it.

Incidentally, I also do not see the 0-255 brightness range being desirable in the API. It could have been 0-100 from the start and brightness_pct would not even have been needed. Same with color_temp; kelvin is what people understand.

Reply:
Cool, I think we're on the same page now. I'd be OK with either dropping RGB or keeping it around in this alternate preprocessed only form. I agree we should make our decision as-if we were designing a new API. Breaking changes suck, but we shouldn't let that stop us from building maintainable code.

I'll try to get this branch up-to-date with what we've discussed, and we can open the API question to others for input once there's better code to talk about.